# Boosting

## The process of giving higher relevance to a set of documents over others is called __boosting__

A good example is google in itself where you are shown some search results boosted to the top (or a place which catches your attention) as they would be from a sponsored source. In essence google has boosted the sponsored search result to the top to bring it to prominence. A well designed search interface would provide the ability to adapt to user input and modify the search results accordingly, say in drilling down or choose from among a first of equals. This is where boosting can play a big part.
There are two stages where documents can be boosted: At index time and at query time.

![alt text](0.png)

## 1. Index time boost

Index time boosting is basically programatically setting the score of a field(s) at the time of indexing. However, you are not actually setting the score here, score is dependent on a lot of factors (for example the tokens in the query in itself which adds to the score), so what is being set is a number against a field which plays a part in the calculation of the score based on the query. This is where __NORM__ comes into play. Norm is basically that one number against the field which affects the document’s score and thus position in the search result pecking order. Norm basically is short for normalized value. The Norm values are added to the index and this can potentially (again, potentially) help increase the query time.
When should I use index time boosting?

_This pretty much depends on the business scenario at hand. For those scenarios where you know which subset of documents need to be boosted before hand, index time boosting would come in useful. Let us take a real world example here, say you have a shopping site selling cars with visitors from around the world. It is required that the search results for cars be boosted to the country of the user currently logged in. Say boost all products which are based in India to those users who have current address country as India?_

To increase the scores for certain documents that match a query, regardless of what that query may be, one can use index-time boosts.
~~~
document.setDocumentBoost(x)
~~~
The default boost for a field is 1, so setting a value between 0 and 1 would down boost the document.

Index-time boosts can be specified per-field also, so only queries matching on that specific field will get the extra boost.

It is also possible to add different boosts to different fields of a document. The only requirement here is that the boosted fields. 
~~~
document.addField(“title”, “Foo Bar”, x);
~~~
It’s important to know that the boost (either for a document or for a field) will be considered when calculating the final score for a document given a search. It is not the final score of the document. Boosting documents is not the same as sorting documents.

## 2. Query time boost
Boosting at query time is much more dynamic as it doesn’t require re-indexing and can be specified with every new request. Also, what gets boosted is not a document or a field, but a __subquery__ on the search. The simplest way to achieve query time boosting is by using the __^__ character plus the boost number on the query, for example:
~~~
foo^5 bar
~~~
Much more complex expressions can also be used for query time boosting, like:
~~~
title:(foo bar)^5 OR content:(foo bar)^2 OR foo OR bar
title:(foo bar)^5 OR title:”foo bar”^20 OR ...
~~~
The syntax can be very simple for simple cases, but it will get more and more complex with more complex use cases.

However, this syntax requires having an expert user who knows how to use it, or some application logic to inject it in the background after the user enters the query and before sending request. Dismax provides other alternatives for query time boosting, as dynamic as the previous one, but with a much easier syntax.

# Boosting a terms
Lucene/Solr provides the relevance level of matching documents based on the terms found. To boost a term use the caret symbol __^__ with a boost factor (a number) at the end of the term you are searching. __The higher the boost factor, the more relevant the term will be.__

Boosting allows you to control the relevance of a document by boosting its term. For example, if you are searching for
`"jakarta apache"` and you want the term `"jakarta"` to be more relevant, you can boost it by adding the `^` symbol along with the boost factor immediately after the term. 

For example, you could type: `jakarta^4 apache`

This will make documents with the term `jakarta` appear more relevant. You can also boost __Phrase Terms__ as in the example:
`"jakarta apache"^4 "Apache Lucene"`

By default, the boost factor is __1__. Although the boost factor must be positive, it can be less than 1 (for example, it could be 0.2).	

# Boosting Fields
The Dismax Query Parser (QP) will create a query that will be executed on many different fields, even if the user hasn’t specified any. This is one of the most important improvements of the Dismax QP over the Lucene QP. But sometimes, not all the fields have the same importance. 

_Sometimes, a hit on the title field is more important than a hit on the content field, or a hit on the content can be more important than a hit on the comments field. _

The Dismax Query Parser provides the ability to consider some fields more important than others with the “qf” (named after “query fields”) parameter, the same that is used for specifying the different fields on which to execute the user query. A common value for this parameter could be:
`qf=title^5 content^2 comments^0.5`
This will translate a user query like “boo bar” into something similar to:
`title:(foo bar)^5 OR content:(foo bar)^2 OR comments:(foo bar)^0.5`

# Boosting Phrases
The same as with query fields, Dismax Query Parser will execute the user query as a phrase query on the specified “phrase” fields. In this parameter, and in a similar way as in the qf parameter, a different boost for each of the phrase fields can be specified:
`pf=title^20 content^10`

This will translate a user query like foo bar into:
`title:”foo bar”^20 OR content:”foo bar”^10`
The last query will only be used for boosting the documents resulting from the original query.

# Boost Queries
Sometimes it is necessary to boost some documents regardless of the user query. A typical example of boost queries is __boosting sponsored documents__. The user searches for “car rental”, but the application has some sponsored document that should be boosted. A good way of doing this is by using boost queries. 

__A boost query is a query that will be executed on background after a user query, and that will boost the documents that matched it.__

For this example, the boost query (specified by the “bq” parameter) would be something like:
bq=sponsored:true
The boost query won’t determine which documents are considered a hit an which are not, but it will just influence the score of the result.

# Boost Functions
Boost Functions are very similar to boost queries; in fact, they can achieve the same goals. The difference between boost functions and boost queries is that the boost function is an arbitrary function instead of a query. 
__A typical example of boost functions is boosting those documents that are more recent than others.__ 
Imagine a forum search application, where the user is searching for forum entries with the text “foo bar”. The application should display all the forum entries that talk about “foo bar” but usually the most recent entries are more important (most users will want to see updated entries, and not historical). 

The boost function will be executed on background after each user query, and will boost some documents in some way.
For this example, a boost function (specified by the “bf” parameter) could be something like:
`bf=recip(ms(NOW,publicationDate),3.16e-11,1,1)`

The same as with the boost queries, this function will not determine which documents are a hit and which are not, it will just add additional score to them.

A note on boost functions: boost functions can also be used with the Lucene QP by using the “_val_” special key inside the query.

# Example

In [1]:
import sys
import lucene
 
from java.nio.file import Paths
from java.util import LinkedHashMap
from org.apache.lucene.analysis.standard import StandardAnalyzer,ClassicAnalyzer
from org.apache.lucene.search import IndexSearcher,BoostQuery,BooleanQuery,TermQuery
from org.apache.lucene.search import BooleanClause
from org.apache.lucene.index import DirectoryReader, Term
from org.apache.lucene.queryparser.classic import QueryParser,MultiFieldQueryParser 
from org.apache.lucene.store import SimpleFSDirectory
from org.apache.lucene.queryparser.simple import SimpleQueryParser
from org.apache.lucene.queryparser.classic import QueryParserBase
from org.apache.lucene.util import Version
import pandas as pd

In [2]:
lucene.initVM()
directory = SimpleFSDirectory(Paths.get('index'))
reader = DirectoryReader.open(directory)
searcher = IndexSearcher(reader)   
analyzer = StandardAnalyzer()
fields = ['title', 'genres', 'directors', 'top_3_cast', 'storyline','synopsis']
clauses = [ BooleanClause.Occur.SHOULD, BooleanClause.Occur.SHOULD, BooleanClause.Occur.SHOULD, \
           BooleanClause.Occur.SHOULD, BooleanClause.Occur.SHOULD, BooleanClause.Occur.SHOULD]
parser = MultiFieldQueryParser(fields, analyzer)
parser.setDefaultOperator(QueryParserBase.OR_OPERATOR)

# Query Time Boosting Terms
# query = 'top_3_cast:"Tom"'

In [3]:
query = 'top_3_cast:"Tom"'
query = MultiFieldQueryParser.parse(parser, query)
hits = searcher.search(query, 100)
print ("Found %d document(s) that matched query '%s':" % (hits.totalHits, query))

Found 302 document(s) that matched query 'top_3_cast:tom':


In [4]:
df = pd.DataFrame()
for hit in hits.scoreDocs:
    #print (hit.score, hit.doc, hit.toString())
    #print("hit.score: ", hit.score)
    #print("hit.doc: ", hit.doc)
    #print("hit: ", hit)
    
    doc = searcher.doc(hit.doc)
    df = df.append([[doc.get('id'), doc.get('title'), doc.get('genres'), doc.get('directors'), doc.get('top_3_cast'), doc.get('storyline'), doc.get('synopsis'), hit.score]], ignore_index = True)
    #print("explain: ", searcher.explain(query,hit.doc).toString())
    #print("________________________")

~~~
explain:  5.541642 = weight(top_3_cast:tom in 5555) [BM25Similarity], 
result of:
  5.541642 = score(doc=5555,freq=2.0 = termFreq=2.0), 
  product of:
    3.9971538 = idf, 
    computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) 
    from:
      302.0 = docFreq
      16468.0 = docCount
    1.386397 = tfNorm, 
    computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) 
    from:
      2.0 = termFreq=2.0
      1.2 = parameter k1
      0.75 = parameter b
      6.1806536 = avgFieldLength
      6.0 = fieldLength
~~~

In [5]:
df.columns = ['id','title', 'genres', 'director', 'top_3_cast', 'storyline', 'synopsis', 'score']
df.head(13)

id                          title  \
0   0120815            Saving Private Ryan   
1   1645187       Walk a Mile in My Pradas   
2   0106218                      Ad Fundum   
3   2461216       William and the Windmill   
4   2124818                        Janapar   
5   0478209  Metal: A Headbanger's Journey   
6   0056868                     Billy Liar   
7   0089945             Rustlers' Rhapsody   
8   0120755         Mission: Impossible II   
9   0240515            Freddy Got Fingered   
10  0404978                         Gambit   
11  0985593                Miss Conception   
12  1172570                        Bronson   

                                     genres  \
0                          ['Drama', 'War']   
1                                ['Comedy']   
2                                 ['Drama']   
3     ['Documentary', 'Biography', 'Drama']   
4   ['Documentary', 'Adventure', 'Romance']   
5                  ['Documentary', 'Music']   
6            ['Comedy', 'Drama', 'Romance']   
7                     ['Comedy', 'Western']   
8       ['Action', 'Adventure', 'Thriller']   
9                                ['Comedy']   
10                      ['Comedy', 'Crime']   
11                    ['Comedy', 'Romance']   
12         ['Action', 'Biography', 'Crime']   

                                             director  \
0                                ['Steven Spielberg']   
1                                  ['Joey Sylvester']   
2                                   ['Erik Van Looy']   
3                                      ['Ben Nabors']   
4                    ['James W. Newton', 'Tom Allen']   
5   ['Sam Dunn', 'Scot McFadyen', '1 more credit',...   
6                                ['John Schlesinger']   
7                                     ['Hugh Wilson']   
8                                        ['John Woo']   
9                                       ['Tom Green']   
10                                ['Michael Hoffman']   
11                                    ['Eric Styles']   
12                           ['Nicolas Winding Refn']   

                                           top_3_cast  \
0     ['Tom Hanks', ' Tom Sizemore', ' Edward Burns']   
1   ['Nathaniel Marston', ' Tom Archdeacon', ' Tom...   
2   ['Tom Van Landuyt', ' Mathias Sercu', ' Tom Va...   
3                ['William Kamkwamba', ' Tom Rielly']   
4                      ['Tom Allen', ' Andrew Welch']   
5     ['Tom Araya', ' Gavin Baddeley', ' Blasphemer']   
6   ['Tom Courtenay', ' Wilfred Pickles', ' Mona W...   
7   ['Tom Berenger', ' G.W. Bailey', ' Marilu Henn...   
8   ['Tom Cruise', ' Dougray Scott', ' Thandie New...   
9      ['Tom Green', ' Rip Torn', ' Marisa Coughlan']   
10  ['Colin Firth', ' Tom Courtenay', ' Alan Rickm...   
11  ['Heather Graham', ' Mia Kirshner', ' Tom Ellis']   
12    ['Tom Hardy', ' Kelly Adams', ' Luing Andrews']   

                                            storyline  \
0       Opening with the Allied invasion of Norman...   
1       A magic Christmas ornament turns two men's...   
2       Sammy Raes, a nice, naive law-freshman fro...   
3       William Kamkwamba, a young Malawian, build...   
4       23-year-old Englishman Tom Allen is all se...   
5       Sam Dunn is a 30-year old anthropologist w...   
6       A young British clerk in a gloomy North Co...   
7       While the audience watches a black and whi...   
8       Chimera is a deadly virus that will bear a...   
9       Gordon, 28, an aspiring animator, leaves h...   
10      Curator Harry Deane is an expert in fine a...   
11      Georgina is an ambitious young London prof...   
12      In 1974, a hot-headed 19 year old named Mi...   

                                             synopsis     score  
0                                                 NaN  5.541642  
1                                                 NaN  5.541642  
2                                                 NaN  5.075861  
3   Saint James Films and TomCat Films, LLC 

# Tom Hanks + Tom Sizemore 0 (score 5.541642) 
# Tom Hardy 12 (score 4.045527)






# query = 'top_3_cast:"Tom" OR top_3_cast:"Tom Hardy"^2'
# 2.0 = boost

In [6]:
query = 'top_3_cast:"Tom" OR top_3_cast:"Tom Hardy"^2'
query = MultiFieldQueryParser.parse(parser, query)
hits = searcher.search(query, 100)
print ("Found %d document(s) that matched query '%s':" % (hits.totalHits, query))

Found 302 document(s) that matched query 'top_3_cast:tom (top_3_cast:"tom hardy")^2.0':


In [7]:
df = pd.DataFrame()
for hit in hits.scoreDocs:
    #print (hit.score, hit.doc, hit.toString())
    #print("hit.score: ", hit.score)
    print("hit.doc: ", hit.doc)
    #print("hit: ", hit)
    
    doc = searcher.doc(hit.doc)
    df = df.append([[doc.get('id'), doc.get('title'), doc.get('genres'), doc.get('directors'), doc.get('top_3_cast'), doc.get('storyline'), doc.get('synopsis'), hit.score]], ignore_index = True)
    print("explain: ", searcher.explain(query,hit.doc).toString())
    print("________________________")

hit.score:  23.387500762939453
hit.doc:  681
explain:  23.3875 = sum of:
  4.045527 = weight(top_3_cast:tom in 681) [BM25Similarity], result of:
    4.045527 = score(doc=681,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) from:
        1.0 = termFreq=1.0
        1.2 = parameter k1
        0.75 = parameter b
        6.1806536 = avgFieldLength
        6.0 = fieldLength
  19.341974 = weight(top_3_cast:"tom hardy" in 681) [BM25Similarity], result of:
    19.341974 = score(doc=681,freq=1.0 = phraseFreq=1.0
), product of:
      2.0 = boost
      9.555349 = idf(), sum of:
        3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
          302.0 = docFreq
          16468.0 = docCount
        5.558195 = idf, 

explain:  4.045527 = sum of:
  4.045527 = weight(top_3_cast:tom in 1792) [BM25Similarity], result of:
    4.045527 = score(doc=1792,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) from:
        1.0 = termFreq=1.0
        1.2 = parameter k1
        0.75 = parameter b
        6.1806536 = avgFieldLength
        6.0 = fieldLength

________________________
hit.score:  4.04552698135376
hit.doc:  1812
explain:  4.045527 = sum of:
  4.045527 = weight(top_3_cast:tom in 1812) [BM25Similarity], result of:
    4.045527 = score(doc=1812,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, compute

explain:  4.045527 = sum of:
  4.045527 = weight(top_3_cast:tom in 3774) [BM25Similarity], result of:
    4.045527 = score(doc=3774,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) from:
        1.0 = termFreq=1.0
        1.2 = parameter k1
        0.75 = parameter b
        6.1806536 = avgFieldLength
        6.0 = fieldLength

________________________
hit.score:  4.04552698135376
hit.doc:  3832
explain:  4.045527 = sum of:
  4.045527 = weight(top_3_cast:tom in 3832) [BM25Similarity], result of:
    4.045527 = score(doc=3832,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, compute

~~~
explain:  23.3875 = sum of:
  4.045527 = weight(top_3_cast:tom in 681) [BM25Similarity], 
  result of:
    4.045527 = score(doc=681,freq=1.0 = termFreq=1.0
), product of:
      3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) 
      from:
        302.0 = docFreq
        16468.0 = docCount
      1.0121019 = tfNorm, computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) from:
        1.0 = termFreq=1.0
        1.2 = parameter k1
        0.75 = parameter b
        6.1806536 = avgFieldLength
        6.0 = fieldLength
  19.341974 = weight(top_3_cast:"tom hardy" in 681) [BM25Similarity], 
  result of:
    19.341974 = score(doc=681,freq=1.0 = phraseFreq=1.0
), product of:
      2.0 = boost
      9.555349 = idf(), sum of:
        3.9971538 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
          302.0 = docFreq
          16468.0 = docCount
        5.558195 = idf, computed as log(1 + (docCount - docFreq + 0.5) / (docFreq + 0.5)) from:
          63.0 = docFreq
          16468.0 = docCount
      1.0121019 = tfNorm, computed as (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * fieldLength / avgFieldLength)) from:
        1.0 = phraseFreq=1.0
        1.2 = parameter k1
        0.75 = parameter b
        6.1806536 = avgFieldLength
        6.0 = fieldLength

~~~

In [8]:
df.columns = ['id','title', 'genres', 'director', 'top_3_cast', 'storyline', 'synopsis', 'score']
df.head(40)

id                                             title  \
0   1172570                                           Bronson   
1   1291584                                           Warrior   
2   1345836                             The Dark Knight Rises   
3   0375912                                        Layer Cake   
4   2692904                                             Locke   
5   1891806                                    From the Ashes   
6   1212450                                           Lawless   
7   1596350                                    This Means War   
8   0120815                               Saving Private Ryan   
9   1645187                          Walk a Mile in My Pradas   
10  0106218                                         Ad Fundum   
11  2461216                          William and the Windmill   
12  2124818                                           Janapar   
13  0478209                     Metal: A Headbanger's Journey   
14  0056868                                        Billy Liar   
15  0089945                                Rustlers' Rhapsody   
16  0120755                            Mission: Impossible II   
17  0240515                               Freddy Got Fingered   
18  0404978                                            Gambit   
19  0985593                                   Miss Conception   
20  1343755                                      Zweiohrküken   
21  1533087                                      Soda Springs   
22  1631867                                  Edge of Tomorrow   
23  1700845                               The Invisible Woman   
24  1843316                       Transformers: The Ride - 3D   
25  1904887                              En fiende att dö för   
26  2483260                                  The Pirate Fairy   
27  0037690                       The Falcon in San Francisco   
28  0064775                                             Otley   
29  0117060                               Mission: Impossible   
30  1179258                                             Chéri   
31  1196204                                 Cemetery Junction   
32  1320239                                    Burke and Hare   
33  1391116                                        Resistance   
34  1483013                                          Oblivion   
35  1756447  Carl Panzram: The Spirit of Hatred and Vengeance   
36  1894476                                    How I Live Now   
37  0060666                          Manos: The Hands of Fate   
38  0085244                                     The Big Chill   
39  0085461                                       The Dresser   

                                     genres  \
0          ['Action', 'Biography', 'Crime']   
1                        ['Drama', 'Sport']   
2                    ['Action', 'Thriller']   
3              ['Action', 'Crime', 'Drama']   
4                                 ['Drama']   
5                  ['Documentary', 'Sport']   
6                        ['Crime', 'Drama']   
7           ['Action', 'Comedy', 'Romance']   
8                          ['Drama', 'War']   
9                                ['Comedy']   
10                                ['Drama']   
11    ['Documentary', 'Biography', 'Drama']   
12  ['Documentary', 'Adventure', 'Romance']   
13                 ['Documentary', 'Music']   
14           ['Comedy', 'Drama', 'Romance']   
15                    ['Comedy', 'Western']   
16      ['Action', 'Adventure', 'Thriller']   
17                               ['Comedy']   
18                      ['Comedy', 'Crime']   
19                    ['Comedy', 'Romance']   
20                    ['Comedy', 'Romance']   
21                                ['Drama']   
22                     ['Action', 'Sci-Fi']   
23        ['Biography', 'Drama', 'History']   
24         ['Short', 'Action', 'Adventure']   
25           ['History', 'Thriller', 'War']   
26     ['Animation', 'Adventure', 'Family']   
27                     ['Cr

# Tom Hardy 0-7 (score 23)
# Tom Hanks 8 (score 5)